substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java

```diff
@@ -556,9 +556,12 @@ public List<String> getBuilderJavaArgs() {
         public List<Path> getBuilderModulePath() {
             List<Path> result = new ArrayList<>();
             // Non-jlinked JDKs need truffle and graal-sdk on the module path since they
-            // don't have those modules as part of the JDK.
+            // don't have those modules as part of the JDK. They also need
+            // word/collection/nativeimage modules on the module path after the split in
+            // GR-43819 (#7171)
             if (libJvmciDir != null) {
                 result.addAll(getJars(libJvmciDir, "graal-sdk", "enterprise-graal"));
+                result.addAll(getJars(libJvmciDir, "word", "collections", "nativeimage"));
```

Comment:
I'm unsure about this one, we need some magic to get `polyglot.jar` copied to the dists dir:

```
Building native-image.image... [dependency SVM_DRIVER updated]
Error occurred during initialization of boot layer
java.lang.module.FindException: Module org.graalvm.polyglot not found

Building native-image.image: Failed due to error: 1
```

```
$ find sdk/mxbuild/linux-amd64/GRAALVM_44571D84E8_JAVA21_STAGE1/graalvm-44571d84e8-java21-23.1.0-dev/lib/jvmci/
sdk/mxbuild/linux-amd64/GRAALVM_44571D84E8_JAVA21_STAGE1/graalvm-44571d84e8-java21-23.1.0-dev/lib/jvmci/
sdk/mxbuild/linux-amd64/GRAALVM_44571D84E8_JAVA21_STAGE1/graalvm-44571d84e8-java21-23.1.0-dev/lib/jvmci/graal.src.zip
sdk/mxbuild/linux-amd64/GRAALVM_44571D84E8_JAVA21_STAGE1/graalvm-44571d84e8-java21-23.1.0-dev/lib/jvmci/collections.jar
sdk/mxbuild/linux-amd64/GRAALVM_44571D84E8_JAVA21_STAGE1/graalvm-44571d84e8-java21-23.1.0-dev/lib/jvmci/collections.src.zip
sdk/mxbuild/linux-amd64/GRAALVM_44571D84E8_JAVA21_STAGE1/graalvm-44571d84e8-java21-23.1.0-dev/lib/jvmci/nativeimage.src.zip
sdk/mxbuild/linux-amd64/GRAALVM_44571D84E8_JAVA21_STAGE1/graalvm-44571d84e8-java21-23.1.0-dev/lib/jvmci/word.jar
sdk/mxbuild/linux-amd64/GRAALVM_44571D84E8_JAVA21_STAGE1/graalvm-44571d84e8-java21-23.1.0-dev/lib/jvmci/graal.jar
sdk/mxbuild/linux-amd64/GRAALVM_44571D84E8_JAVA21_STAGE1/graalvm-44571d84e8-java21-23.1.0-dev/lib/jvmci/graal-management.src.zip
sdk/mxbuild/linux-amd64/GRAALVM_44571D84E8_JAVA21_STAGE1/graalvm-44571d84e8-java21-23.1.0-dev/lib/jvmci/graal-management.jar
sdk/mxbuild/linux-amd64/GRAALVM_44571D84E8_JAVA21_STAGE1/graalvm-44571d84e8-java21-23.1.0-dev/lib/jvmci/word.src.zip
sdk/mxbuild/linux-amd64/GRAALVM_44571D84E8_JAVA21_STAGE1/graalvm-44571d84e8-java21-23.1.0-dev/lib/jvmci/nativeimage.jar
```

Reply:
What's the command you are using to build GraalVM in this case? If `polyglot` is not part of the dependencies then not copying `polyglot.jar` over is correct.

E.g.

```
mx --primary-suite-path substratevm --no-jlinking --components="Native Image,LibGraal" --native-images=native-image,lib:jvmcicompiler build
```

Shouldn't copy `polyglot.jar`.

On the other hand:

```
mx --primary-suite-path substratevm --no-jlinking --components="Native Image,LibGraal,poly" --native-images=native-image,lib:jvmcicompiler build
```

Should probably copy it over.

Reply:
> Any idea on how to achieve that?

I am not sure what you refer to with _that_.

I think the reason you are getting:

```
Building native-image.image... [dependency SVM_DRIVER updated]
Error occurred during initialization of boot layer
java.lang.module.FindException: Module org.graalvm.polyglot not found
```

Is because we explicitly pass `--add-modules org.graalvm.polyglot` to the command line despite `polyglot.jar` not being present (or needed).

Are you getting this when you run `getJars(libJvmciDir, "polyglot"` as well?

Reply:
"That" being making `mx --primary-suite-path substratevm --no-jlinking --components="Native Image,LibGraal" --native-images=native-image,lib:jvmcicompiler build` pass. It looks like the build invokes the native-image launcher macro, which in-turns fails to boot because of the missing `polyglot.jar` getting copied to the sdk's image directory and the launcher script referencing it using `--add-modules`.

> Are you getting this when you run `getJars(libJvmciDir, "polyglot"` as well?

Yes.

Reply:
> and the launcher script referencing it using `--add-modules`.

Exactly, that's the issue. That's why I believe we should remove it from the bash/native launcher and have `NativeImage` add it only when necessary as I think the launcher doesn't really need it...

Reply:
OK. I don't know why `--add-modules ...polyglot` was added and have, therefore, no way of testing it. Do you have ideas, suggestions what executes that codepath?

Reply:
Unfortunately not, @chumer could you please shed some light on this?